src/EventListener/BulkSavePropertyListener.php

```diff
@@ -0,0 +1,105 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Flagbit\Bundle\CategoryBundle\EventListener;
+
+use Akeneo\Pim\Enrichment\Component\Category\Model\CategoryInterface;
+use Doctrine\ORM\EntityManagerInterface;
+use Flagbit\Bundle\CategoryBundle\Entity\CategoryProperty;
+use Flagbit\Bundle\CategoryBundle\Repository\CategoryPropertyRepository;
+use Symfony\Component\EventDispatcher\GenericEvent;
+use Symfony\Component\HttpFoundation\ParameterBag;
+
+use function count;
+use function is_array;
+
+/**
+ * Category post save listener that handles bulk saves with properties.
+ */
+class BulkSavePropertyListener
+{
+    /** @var ParameterBag<string, array<string, mixed>> */
+    private ParameterBag $propertiesBag;
+    private CategoryPropertyRepository $repository;
+    private EntityManagerInterface $entityManager;
+
+    /**
+     * @phpstan-param ParameterBag<string, array<string, mixed>> $propertiesBag
+     */
+    public function __construct(
+        ParameterBag $propertiesBag,
+        CategoryPropertyRepository $repository,
+        EntityManagerInterface $entityManager
+    ) {
+        $this->propertiesBag = $propertiesBag;
+        $this->repository    = $repository;
+        $this->entityManager = $entityManager;
+    }
+
+    /**
+     * @phpstan-param GenericEvent<mixed> $event
+     */
+    public function onBulkCategoryPostSave(GenericEvent $event): void
+    {
+        // Save category properties
+        foreach ($event->getSubject() as $category) {
+            if (! $category instanceof CategoryInterface) {
+                continue;
+            }
+
+            if (! $this->propertiesBag->has($category->getCode())) {
+                continue;
+            }
+
+            $properties = $this->propertiesBag->get($category->getCode());
+            if (count($properties) === 0) {
+                return;
+            }
+
+            $categoryProperty = $this->findProperty($category);
+            $categoryProperty->setProperties($this->mergeProperties($categoryProperty->getProperties(), $properties));
+
+            $this->entityManager->persist($categoryProperty);
+            $this->entityManager->flush();
```

Comment:
Flush inside a loop is a performance issue, but to prevent data loss on error we will keep it that way. Let's see how big the impact is in a project.

Reply:
I also considered moving the flush out of the loop during development but moved it back to ensure not all data will be lost on error. I think as most PIM users will have way more products (approx. a few tens or hundreds of categories vs thousands of products) than categories, the performance impact shouldn't be too bad.

Maybe also a try-catch around this entire thing would be good to just log errors and continue with the left over properties?

Reply:
> Maybe also a try-catch around the loops entire content would be good to just log errors and continue with the left over properties? This way it would behave a bit more like Akeneo's own imports, which just skip broken items.

I'd only expect critical errors / fatal errors in that loop. Those should result in breaking and stopping an import.